### PR TITLE
Add common library pitfalls

### PR DESCRIPTION
**Pull request by pipermerriam**
Added two pitfalls I have experienced during contract development.  @chriseth can you please confirm that these are both _real_ things, as well as my wording being precise enough?
#### Cute animal picture

animal-costumes-1

### CODE REVIEW

File: _docs/contracts.md [hunk 1]
```diff
@@ -113,6 +113,11 @@ Restrictions for libraries in comparison to contracts:
 
 (these might be lifted at a later point)
 
+Common pitfalls for libraries
+
+* The value for `msg.sender` will be that of the contract which is calling the library function.  (If A calls contract B which internally calls library C, then within the function call of library C, `msg.sender` will be the address of contract B).
+* When declaring local variables for structs or arrays, modifications to them will not persist in storage unless they are declared with `storage` as their storage location.  (For example, `var my_struct = self.some_mapping[key]` any changes to `my_struct` would not be persisted in contract storage).
```
Comment (chriseth) [inline _docs/contracts.md]: I think this should rather go to the "reference types" section, but we can add a link here.

What would fit better here would be to say that when handing over a memory type to a library function, modifications will not be carried back to the value in the context of the caller. The reason is the same as above.

File: _docs/contracts.md [hunk 2]
```diff
@@ -113,6 +113,11 @@ Restrictions for libraries in comparison to contracts:
 
 (these might be lifted at a later point)
 
+Common pitfalls for libraries
+
+* The value for `msg.sender` will be that of the contract which is calling the library function.  (If A calls contract B which internally calls library C, then within the function call of library C, `msg.sender` will be the address of contract B).
```
Comment (chriseth) [inline _docs/contracts.md]: I would say that it is important to explain the reason for that here:

The expression `LibraryName.functionName()` performs an external function call, which maps to a real EVM call just like `otherContract.functionName()` or `this.functionName()` and thus extends the call depth by one (limited to 1024), stores the caller as `msg.sender`, it is possible to transfer ether together with this call (using `.value(x)`), creates a completely new memory context, and thus memory types will be copied and cannot be passed by reference, and so on. The semantics of that are planned to be explained in the yet to be written page about the Ethereum Virtual Machine.

File: _docs/contracts.md (hunk 1 above)
Comment (pipermerriam) [inline _docs/contracts.md]: Would you be ok leaving a small note about it here and linking to the "reference types" section for more detail?  This was something that bit me more than once thus far.

Comment (pipermerriam) [inline _docs/contracts.md]: Scratch that.  I've gone ahead and removed this section.  Will look at whether there is something to be added to the reference types section with respect to this.

File: _docs/contracts.md (hunk 2 above)
Comment (pipermerriam) [inline _docs/contracts.md]: Please take a look at my revision.  I took your paragraph and reworded it a bit to better fit the context.

**Pull request merged**